docs/architecture/grpc-for-wcf-developers/types-of-rpc.md

```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
 
 # Types of RPC
 
+As a WCF developer you are probably used to dealing with the following types of RPC:
+
+- Request/Reply
+- Duplex:
+  - One way duplex with session
+  - Full duplex with session
+- One-way
+
+It is possible to map these RPC types fairly naturally to existing gRPC concepts and this chapter will look at each of these areas in turn. Similar examples will be explored in much greater depth in [Chapter 5](migrating-wcf-to-grpc.md).
+
+| WCF | gRPC |
+| --- | ---- |
+| Regular request/reply | Unary |
+| Duplex service with session using a client callback interface | Server streaming |
+| Full duplex service with session | Bi-directional streaming |
+| One-way operations | Client streaming |
+
+## Request/reply
+
+For simple request/reply methods that take and return small amounts of data, use the simplest gRPC pattern, the unary RPC.
+
+```protobuf
+service Things {
+    rpc Get(GetThingRequest) returns (GetThingResponse);
+}
+```
+
+```csharp
+public class ThingService : Things.ThingsBase
+{
+    public override async Task<GetThingResponse> Get(GetThingRequest request, ServerCallContext context)
+    {
+        // Get thing from database
+        return new GetThingResponse { Thing = thing };
+    }
+}
+```
+
+```csharp
+public async Task ShowThing(int thingId)
+{
+    var thing = await _thingsClient.GetAsync(new GetThingRequest { ThingId = thingId });
+    Console.WriteLine($"{thing.Name}");
+}
+```
+
+As you can see, implementing a gRPC unary RPC service method is very similar to implementing a WCF operation, except that with gRPC you override a base class method instead of implementing an interface. Note that on the server, gRPC base methods always return a `Task<T>`, although the client provides both async and blocking methods to call the service.
+
+## WCF duplex, one-way to client
+
+WCF applications (with certain bindings) can create a persistent connection between client and server, and the server can asynchronously send data to the client until the connection is closed, using a *callback interface*.
+
+gRPC services can work with streams of messages, in either or both directions. This does not map exactly to WCF Duplex, but the same result can be achieved.
+
+Here is an example of a server streaming RPC.
+
+```protobuf
+service ThingStreamer {
+    rpc Subscribe(SubscribeRequest) returns (stream Thing);
+}
+```
+
+```csharp
+{
+    public class ThingStreamerService : ThingStreamer.ThingStreamerBase
+    {
+        public override async Task Subscribe(SubscribeRequest request,
+            IServerStreamWriter<Thing> responseStream,
+            ServerCallContext context)
+        {
+            while (!context.CancellationToken.IsCancellationRequested)
+            {
+                var thing = await _things.GetNext();
```

Comment:
Where does _things come from? IMO simplify with declaring an int outside the loop and increase the returned string:

```cs
await responseStream.WriteAsync(new Thing { name = $"Thing {i}" });
```